File: engine/src/investigation/tools-read.ts
import * as fs from "node:fs";
import * as path from "node:path";

const MAX_FILE_READ = 100_000; // 100 KB
const MAX_SEARCH_RESULTS = 50;
const CONTEXT_LINES = 3;
const SKIP_DIRS = new Set(["node_modules", ".git", ".svn"]);
const TEXT_EXTS = new Set([".js", ".mjs", ".cjs", ".ts", ".mts", ".json", ".md", ".txt", ".yml", ".yaml"]);

function safePath(packagePath: string, relPath: string): string | null {
  const abs = path.normalize(path.join(packagePath, relPath));
  const base = path.normalize(packagePath);
  if (!abs.startsWith(base + path.sep) && abs !== base) return null;
  return abs;
}

export function readFileImpl(packagePath: string, relPath: string): string {
  const abs = safePath(packagePath, relPath);
  if (!abs) return `ERROR: path traversal blocked: ${relPath}`;
  if (!fs.existsSync(abs) || !fs.statSync(abs).isFile()) {
    return `ERROR: file not found: ${relPath}`;
  }
  try {
    const size = fs.statSync(abs).size;
    if (size > MAX_FILE_READ) return `ERROR: file too large (${size} bytes, max ${MAX_FILE_READ})`;
    return fs.readFileSync(abs, "utf-8");
  } catch (err) {
    return `ERROR: ${err}`;
  }
}

export function listFilesImpl(packagePath: string): string {
  const entries: Array<{ path: string; size: number; ext: string | null }> = [];

  function walk(dir: string) {
    let dirents: fs.Dirent[];
    try {
      dirents = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of dirents) {
      const abs = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(entry.name)) walk(abs);
        continue;
      }
      if (!entry.isFile()) continue;
      let size = -1;
      try { size = fs.statSync(abs).size; } catch { /* skip */ }
      const ext = path.extname(entry.name) || null;
      entries.push({ path: path.relative(packagePath, abs), size, ext });
    }
  }

  walk(packagePath);
  return JSON.stringify(entries, null, 2);
}

export function searchFilesImpl(packagePath: string, pattern: string): string {
  let regex: RegExp;
  try {
    regex = new RegExp(pattern, "i");
  } catch (err) {
    return `ERROR: invalid regex: ${err}`;
  }

  const results: string[] = [];

  function walk(dir: string) {
    let dirents: fs.Dirent[];
    try {
      dirents = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of dirents) {
      const abs = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(entry.name)) walk(abs);
        continue;
      }
      if (!entry.isFile()) continue;
      const ext = path.extname(entry.name);
      if (!TEXT_EXTS.has(ext)) continue;

      let lines: string[];
      try {
        lines = fs.readFileSync(abs, "utf-8").split("\n");
      } catch {
        continue;
      }

      const rel = path.relative(packagePath, abs);
      for (let i = 0; i < lines.length; i++) {
        if (regex.test(lines[i])) {
          const start = Math.max(0, i - CONTEXT_LINES);
          const end = Math.min(lines.length, i + CONTEXT_LINES + 1);
          const snippet = lines
            .slice(start, end)
            .map((l, j) => `  ${j + start === i ? ">" : " "} ${j + start + 1}: ${l}`)
            .join("\n");
          results.push(`[${rel}:${i + 1}]\n${snippet}`);

          if (results.length >= MAX_SEARCH_RESULTS) {
            results.push(`... truncated at ${MAX_SEARCH_RESULTS} results`);
            return;
          }
        }
      }
    }
  }

  walk(packagePath);
  return results.length ? results.join("\n") : `No matches for pattern: ${pattern}`;
}

File: engine/src/npmguard/investigation/tools_read.py
"""Host-side read tools for the investigation agent.

These operate directly on the extracted package tarball on the host filesystem.
No Docker needed. All paths are validated to stay within the package directory.
"""

from __future__ import annotations

import json
import os
import re

import structlog

log = structlog.get_logger()

MAX_FILE_READ = 100_000  # 100 KB
MAX_SEARCH_RESULTS = 50
CONTEXT_LINES = 3


def _safe_path(package_path: str, rel_path: str) -> str | None:
    """Resolve *rel_path* within *package_path*, rejecting traversal attempts."""
    abs_path = os.path.normpath(os.path.join(package_path, rel_path))
    if not abs_path.startswith(os.path.normpath(package_path) + os.sep) and abs_path != os.path.normpath(package_path):
        return None
    return abs_path


def read_file(package_path: str, path: str) -> str:
    """Read a file from the package. Returns content or an error string."""
    abs_path = _safe_path(package_path, path)
    if abs_path is None:
        return f"ERROR: path traversal blocked: {path!r}"

    if not os.path.isfile(abs_path):
        return f"ERROR: file not found: {path!r}"

    try:
        size = os.path.getsize(abs_path)
        if size > MAX_FILE_READ:
            return f"ERROR: file too large ({size} bytes, max {MAX_FILE_READ})"
        with open(abs_path, encoding="utf-8", errors="replace") as f:
            return f.read()
    except OSError as exc:
        return f"ERROR: {exc}"


def list_files(package_path: str) -> str:
    """List all files in the package with sizes and extensions. Returns JSON."""
    entries = []
    skip = {"node_modules", ".git", ".svn"}

    for dirpath, dirnames, filenames in os.walk(package_path):
        dirnames[:] = [d for d in dirnames if d not in skip]
        for fname in filenames:
            abs_path = os.path.join(dirpath, fname)
            rel_path = os.path.relpath(abs_path, package_path)
            try:
                size = os.path.getsize(abs_path)
            except OSError:
                size = -1
            _, ext = os.path.splitext(fname)
            entries.append({"path": rel_path, "size": size, "ext": ext or None})

    return json.dumps(entries, indent=2)


def search_files(package_path: str, pattern: str) -> str:
    """Regex search across all text files. Returns matches with context lines."""
    try:
        regex = re.compile(pattern, re.IGNORECASE)
    except re.error as exc:
        return f"ERROR: invalid regex: {exc}"

    skip = {"node_modules", ".git", ".svn"}
    text_exts = {".js", ".mjs", ".cjs", ".ts", ".mts", ".json", ".md", ".txt", ".yml", ".yaml"}
    results: list[str] = []

    for dirpath, dirnames, filenames in os.walk(package_path):
        dirnames[:] = [d for d in dirnames if d not in skip]
        for fname in filenames:
            _, ext = os.path.splitext(fname)
            if ext not in text_exts:
                continue

            abs_path = os.path.join(dirpath, fname)
            rel_path = os.path.relpath(abs_path, package_path)

            try:
                with open(abs_path, encoding="utf-8", errors="replace") as f:
                    lines = f.readlines()
            except OSError:
                continue

            for i, line in enumerate(lines):
                if regex.search(line):
                    start = max(0, i - CONTEXT_LINES)
                    end = min(len(lines), i + CONTEXT_LINES + 1)
                    snippet = "".join(
                        f"  {'>' if j == i else ' '} {j + 1}: {lines[j]}"
                        for j in range(start, end)
                    )
                    results.append(f"[{rel_path}:{i + 1}]\n{snippet}")

                    if len(results) >= MAX_SEARCH_RESULTS:
                        results.append(f"... truncated at {MAX_SEARCH_RESULTS} results")
                        return "\n".join(results)

    if not results:
        return f"No matches for pattern: {pattern!r}"
    return "\n".join(results)
